data/scripts.js

```diff
@@ -292,11 +255,291 @@ let BattleScripts = {
 			return false;
 		}
 
+		return true;
+	},
+	trySpreadMoveHit(targets, pokemon, move) {
+		if (!targets.length) {
+			this.attrLastMove('[notarget]');
+			this.add(this.gen >= 5 ? '-fail' : '-notarget', pokemon);
+			return false;
+		}
+		if (targets.length > 1) move.spreadHit = true;
+
+		/** @type {((targets: Pokemon[], pokemon: Pokemon, move: ActiveMove) => (number | boolean | "" | undefined)[] | undefined)[]} */
+		let moveSteps = [
+			// 0. check for semi invulnerability
+			this.tryImmunityEvent,
+
+			// 1. check for type immunity (this is step 2 in gen 7+)
+			this.typeImmunity,
+
+			// 2. run the 'TryHit' event (Protect, Magic Bounce, Volt Absorb, etc.) (this is step 1 in gen 7)
+			this.tryHitEvent,
+
+			// 3. check for powder immunity
+			this.powderImmunity,
+
+			// 4. check for prankster immunity
+			this.pranksterImmunity,
+
+			// 5. check accuracy
+			this.accuracy,
+
+			// 6. break protection effects
+			this.breakProtect,
+
+			// 7. steal positive boosts (Spectral Thief)
+			this.stealBoosts,
+
+			// 8. loop that processes each hit of the move (has its own steps per iteration)
+			this.moveHitLoop,
+
+			// 9. effects that run after secondary effects, such as Color Change
+			this.afterMoveSecondaryEvent,
+		];
+		if (this.gen >= 7) {
+			// Swap step 1 with step 2
+			[moveSteps[1], moveSteps[2]] = [moveSteps[2], moveSteps[1]];
+		}
+
+		this.setActiveMove(move, pokemon, targets[0]);
+		move.zBrokeProtect = false;
+
+		let hitResult = this.singleEvent('PrepareHit', move, {}, targets[0], pokemon, move);
+		if (!hitResult) {
+			if (hitResult === false) {
+				this.add('-fail', pokemon);
+				this.attrLastMove('[still]');
+			}
+			return false;
+		}
+		this.runEvent('PrepareHit', pokemon, targets[0], move);
+
+		if (!this.singleEvent('Try', move, null, pokemon, targets[0], move)) {
+			return false;
+		}
+
+		let finalResult;
+		let hitResults;
+		for (const step of moveSteps) {
+			hitResults = step.call(this, targets, pokemon, move);
+			if (!hitResults) continue;
+			for (let i = 0; i < targets.length; i++) {
+				if (!hitResults[i]) {
+					// hit resolved for this target.
+					// remove target and corresponding hitResults entry from their respective arrays
+					targets.splice(i, 1);
+					let targetResult = hitResults.splice(i, 1)[0];
+					i--;
+					// store failure/NOT_FAILURE in finalResult, with the following priority:
+					// truthy/number > false > '' > undefined
+					if (targetResult || targetResult === 0) {
+						finalResult = targetResult;
+					} else if (finalResult !== false || targetResult !== this.NOT_FAILURE) {
+						finalResult = targetResult;
+					}
+				}
+			}
+		}
+
+		let moveResult = !!targets.length;
+		if (!moveResult && finalResult === this.NOT_FAILURE) pokemon.moveThisTurnResult = null;
```

Comment:
After reading this code, I believe it's mostly useless. `finalResult` isn't actually used except to check for equality with `this.NOT_FAILURE`, but the code above ensures that `finalResult` will never be `this.NOT_FAILURE`.

Reply:
the problem is that it should be storing `this.NOT_FAILURE` sometimes

Reply:
...and it does. The `finalResult !== false` condition allows that possibility.

Reply:
Well, the conditional is still wrong. I should point out that the reason for `finalResult` is so that we can check for a move (spread move or single target) that only hits a Protect-like, which doesn't count as a failure for Stomping Tantrum, but if it actually fails against even one target, that takes precedence, hence the `truthy/number > false > this.NOT_FAILURE > undefined` hierarchy outlined in the comment. However, since you've pointed out that I'm checking for move failure a different way, I can simplify that heirarchy to `truthy/number/false > this.NOT_FAILURE > undefined`.

Reply:
You should probably just make a boolean `atLeastOneFailure` which you can manually track. That doesn't seem too hard.